File: 1306-JumpGameIii/1306-JumpGameIii.cpp
// Last updated: 9/6/2026, 3:57:39 PM
/* صلِّ على النبي */
#include <bits/stdc++.h>
using namespace std;
// #define int long long
#define sz(s) (int)(s).size()
#define all(s) s.begin(), s.end()

class Solution
{
private:
    // اللهم اهدنا يا رب اليك
    public:
        bool canReach(vector<int>& arr, int start) {
            queue<int> q;
            vector<bool> visited(arr.size()+3, 0);
            q.push(start);
            while(!q.empty()){
                auto x = q.front();
                q.pop();
                visited[x] = 1;
                if(arr[x] == 0) return 1;
                int y = x-arr[x];
                int z = x+arr[x];
                if(y>=0 && !visited[y]) 
                q.push(y);
                if(z < arr.size() && !visited[z]) 
                q.push(z);
            }
        return 0;
    }
};

File: 1306-JumpGameIii/1306-JumpGameIii.java
// Last updated: 9/6/2026, 3:57:19 PM
import java.util.Queue;
import java.util.concurrent.ConcurrentLinkedQueue;

class Solution {
    public boolean canReach(int[] arr, int start) {
        Queue<Integer> q = new ConcurrentLinkedQueue<>();
        boolean visited[] = new boolean[arr.length];
        q.add(start);
        while(!q.isEmpty()){
            Integer x = q.peek();
            q.poll();
            visited[x] = true;
            if(arr[x] == 0) return true;
            Integer y = x - arr[x];
            Integer z = x + arr[x];
            if(y >=0 && !visited[y]) q.add(y);
            if(z < arr.length && !visited[z]) q.add(z);
        }
        return false;
    }
}